akka-stream/src/main/scala/akka/stream/stage/GraphStage.scala

```diff
@@ -215,27 +217,37 @@ object GraphStageLogic {
     }
 
     private val callback = getAsyncCallback(internalReceive)
-    private def cell = materializer.supervisor match {
-      case ref: LocalActorRef                        => ref.underlying
-      case ref: RepointableActorRef if ref.isStarted => ref.underlying.asInstanceOf[ActorCell]
-      case unknown =>
-        throw new IllegalStateException(s"Stream supervisor must be a local actor, was [${unknown.getClass.getName}]")
-    }
 
-    private val functionRef: FunctionRef =
-      cell.addFunctionRef(
-        {
-          case (r, PoisonPill) if poisonPillFallback ⇒
-            callback.invoke((r, PoisonPill))
-          case (_, m @ (PoisonPill | Kill)) =>
-            materializer.logger.warning(
-              "{} message sent to StageActor({}) will be ignored, since it is not a real Actor." +
-              "Use a custom message type to communicate with it instead.",
-              m,
-              functionRef.path)
-          case pair => callback.invoke(pair)
-        },
-        name)
+    private val functionRef: FunctionRef = {
+      val f: (ActorRef, Any) => Unit = {
+        case (r, PoisonPill) if poisonPillFallback ⇒
+          callback.invoke((r, PoisonPill))
+        case (_, m @ (PoisonPill | Kill)) =>
+          materializer.logger.warning(
+            "{} message sent to StageActor({}) will be ignored, since it is not a real Actor." +
+            "Use a custom message type to communicate with it instead.",
+            m,
+            functionRef.path)
+        case pair => callback.invoke(pair)
+      }
+
+      materializer.supervisor match {
+        case ref: LocalActorRef =>
+          ref.underlying.addFunctionRef(f, name)
+        case ref: RepointableActorRef =>
+          if (ref.isStarted)
+            ref.underlying.asInstanceOf[ActorCell].addFunctionRef(f, name)
+          else {
+            // this may happen if materialized immediately before Materializer has been fully initialized,
+            // should be rare
+            implicit val timeout = ref.system.settings.CreationTimeout
+            val reply = (materializer.supervisor ? StreamSupervisor.AddFunctionRef(f, name)).mapTo[FunctionRef]
+            Await.result(reply, timeout.duration)
```

Comment:
How is a deadlock prevented here? Wouldn't that warrant at least a comment?

Reply:
An even simpler reproducer as the one added above indeed deadlocks if run on a small enough thread pool:

```scala
package akka.stream

import akka.actor.ActorRef
import akka.stream.scaladsl.Keep
import akka.stream.scaladsl.Sink
import akka.stream.scaladsl.Source
import akka.stream.testkit.StreamSpec
import org.reactivestreams.Publisher

import scala.concurrent.Await
import scala.concurrent.Future
import scala.concurrent.duration._

class PotentialStreamDeadlock extends StreamSpec(
  """
     akka.actor.creation-timeout = 3s
     akka.actor.default-dispatcher.executor = "thread-pool-executor"
     akka.actor.default-dispatcher.thread-pool-executor.fixed-pool-size = 1
     akka.stream.materializer.dispatcher = "akka.actor.default-dispatcher"
     akka.actor.default-mailbox.mailbox-type = "akka.dispatch.UnboundedMailbox"
  """
){
  "be possible to run immediately, reproducer of #26714" in {
    val res =
    Future {
      val mat = ActorMaterializer()
      val source: Source[String, ActorRef] = Source.actorRef[String](10000, OverflowStrategy.fail)
      val (_: ActorRef, _: Publisher[String]) =
        source.toMat(Sink.asPublisher(false))(Keep.both).run()(mat)
      mat.shutdown()
    }(system.dispatcher)
    Await.result(res, 10.seconds)
  }
}
```

For simple cases, and default configuration it will probably be prevented by idle pool threads and FJP creating spare threads when `Await` is used (and in 2.6 probably by the internal thread pool). However, shouldn't we follow our own mantra and don't block on principle and not depend on circumstances to prevent crashes / deadlocks? (We discussed introducing `Await`s into stream logic when discussion how to remove `Future` from StreamRefs and concluded that we shouldn't do it...)

Reply:
Filed https://github.com/akka/akka/issues/26850 to discuss potential solutions.